src/main/java/ca/spottedleaf/starlight/mixin/common/lightengine/LevelLightEngineMixin.java

```diff
@@ -57,7 +58,12 @@ public final StarLightInterface getLightEngine() {
     )
     public void construct(final LightChunkGetter chunkProvider, final boolean hasBlockLight, final boolean hasSkyLight,
                           final CallbackInfo ci) {
-        this.lightEngine = new StarLightInterface(chunkProvider, hasSkyLight, hasBlockLight, (LevelLightEngine)(Object)this);
+        // avoid ClassCastException in cases where custom LightChunkGetters do not return a Level from getLevel()
+        if (chunkProvider.getLevel() instanceof Level) {
+            this.lightEngine = new StarLightInterface(chunkProvider, hasSkyLight, hasBlockLight, (LevelLightEngine) (Object) this);
+        } else {
+            this.lightEngine = new StarLightInterface(null, hasSkyLight, hasBlockLight, (LevelLightEngine) (Object) this);
+        }
```

Comment:
I'd recommend moving the `instanceof` check into the `StarLightInterface` constructor.

Reply:
That was the first fix I attempted, which resulted in a hang on world load.
![image](https://user-images.githubusercontent.com/3160746/147422246-c7a26d0b-abec-44d2-b46d-e6d017f213f6.png)


Performing the check in LeveLightEngineMixin fixed everything without issue.